### PR TITLE
Add Isolate::set_wasm_streaming_callback()

### LINKED ISSUE #556: Expose wasi_streaming_callback
In Deno, our implementation of `WebAssembly.instantiateStreaming` is broken because I thought the implementation provided by wasi-test-streaming would be runtime API compatible with any implementation of Response.

Turns out, from what I can tell in the implementation found here https://github.com/v8/v8/blob/4b9b23521e6fd42373ebbcb20ebe03bf445494f9/src/wasm/wasm-js.cc#L585 that we're supposed to override wasi_streaming_callback defined here https://github.com/v8/v8/blob/4b9b23521e6fd42373ebbcb20ebe03bf445494f9/src/isolate.h#L372

So for rusty_v8 we need Isolate to expose this callback.

### PR DESCRIPTION
**Pull request by bnoordhuis**
Add the hook necessary to implement `WebAssembly.compileStreaming()`.

Fixes #556

cc @caspervonb

### CODE REVIEW

File: src/binding.cc [hunk 1]
```diff
@@ -2029,6 +2034,38 @@ MaybeBool v8__Module__SetSyntheticModuleExport(const v8::Module& self,
       isolate, ptr_to_local(export_name), ptr_to_local(export_value)));
 }
 
+struct WasmStreamingSharedPtr {
+  std::shared_ptr<v8::WasmStreaming> inner;
+};
+
+static_assert(sizeof(WasmStreamingSharedPtr) <= 2 * sizeof(void*),
```
Comment (ry) [inline src/binding.cc:2041]: Why <= ?

File: src/wasm.rs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    let zero = null_mut();
+    let mut that = WasmStreamingSharedPtr([zero, zero]);
+    unsafe {
+      v8__WasmStreaming__Unpack(scope.get_isolate_ptr(), data, &mut that)
+    };
+    let source = args.get(0);
+    (F::get())(scope, source, WasmStreaming(that));
+  }
+  c_fn::<F>
+}
```
Comment (ry) [inline src/wasm.rs:86]: nice!

File: tests/test_api.rs [hunk 3]
```diff
@@ -4326,3 +4327,68 @@ fn clear_kept_objects() {
...
+  eval(scope, script).unwrap();
+
+  let global = context.global(scope);
+  let name = v8::String::new(scope, "result").unwrap().into();
+  assert!(global.get(scope, name).unwrap().is_null());
+
+  let mut ws = WS.with(|slot| slot.borrow_mut().take().unwrap());
+  assert!(global.get(scope, name).unwrap().is_null());
+
+  ws.on_bytes_received(&[0x00, 0x61, 0x73, 0x6d, 0x01, 0x00, 0x00, 0x00]);
```
Comment (ry) [inline tests/test_api.rs:4368]: What is this wasm code - maybe put a comment with the WAT?

**Review by ry — APPROVED** [review]
LGTM

File: src/binding.cc (hunk 1 above)
Comment (bnoordhuis) [reply src/binding.cc:2041]: `std::shared<T>` is normally two pointers big but it's conceivable that an alternative libc++ might implement it as a single pointer. IOW, over-allocating is okay, under-allocating is not.

File: tests/test_api.rs (hunk 3 above)
Comment (bnoordhuis) [reply tests/test_api.rs:4368]: It doesn't contain any actual WASM bytecode, just the file magic and WASM version. :-) I'll add a comment.

File: src/binding.cc [hunk 4]
```diff
@@ -2029,6 +2034,38 @@ MaybeBool v8__Module__SetSyntheticModuleExport(const v8::Module& self,
       isolate, ptr_to_local(export_name), ptr_to_local(export_value)));
 }
 
+struct WasmStreamingSharedPtr {
```
Comment (piscisaureus) [inline src/binding.cc:2037]: We do not usually wrap a shared_ptr in a struct. Why is that necessary here?

**Comment by piscisaureus** [comment]
I don't understand why the existing infrastructure to wrap C++ `shared_ptr` can't be used here, but I'll admit that there might be a good reason.

File: src/binding.cc (hunk 4 above)
Comment (bnoordhuis) [reply src/binding.cc:2037]: It's not strictly necessary but I like the newtype idiom. It makes it harder to use it wrong.

It's also nice to have the same type name in C++ and Rust, easier to grep.

**Comment by bnoordhuis** [comment]
Explained in the comment here :-)

https://github.com/denoland/rusty_v8/pull/560/files#diff-6d7bc9e99526e7f06b818a6103bd30aaa067263cc8e34a14197e1a3bcf2dd4dbR14-R21
